In [1]:
#loading modules
import numpy as np
import pandas as pd

In [2]:
#Loding Datasets
Transcations = pd.read_excel('Data.xlsx', sheet_name=1, header=1)
NewCustomerList = pd.read_excel('Data.xlsx', sheet_name=2, header=1, dtype={'DOB': 'datetime64[ns]'})
CustomerDemographic = pd.read_excel('Data.xlsx', sheet_name=3, header=1, dtype={'DOB': 'datetime64[ns]'})
CustomerAddress = pd.read_excel('Data.xlsx', sheet_name=4, header=1)

In [3]:
#view first 5 rows of demographic data
CustomerDemographic.head(5)

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [4]:
#checking the number of unique values of the gender coloumn
CustomerDemographic.gender.nunique()

6

In [5]:
#checking the unique values of the unique column
CustomerDemographic.gender.unique()

array(['F', 'Male', 'Female', 'U', 'Femal', 'M'], dtype=object)

**CustomerDemographic**
1. 'gender' column which is suppose to be categorical with choices between male and female has 6 unique values namely: F, Male, Femake, U, Femal , M and blanks

2. 'default' column seems to be a useless chunk of data and also has no relation to the goal set out to be met

In [6]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

**CustomerDemographic**

3. a. There are missing values in 'last_name', 'DOB', 'job_title', 'job_industry_category', 'default' and 'tenure' columns

In [7]:
null_count = []
null_count = CustomerDemographic['last_name'].isnull().sum(), CustomerDemographic['DOB'].isnull().sum(), CustomerDemographic['job_title'].isnull().sum(), CustomerDemographic['job_industry_category'].isnull().sum(), CustomerDemographic['default'].isnull().sum(), CustomerDemographic['tenure'].isnull().sum()
null_count

(125, 87, 506, 656, 302, 87)

**CustomerDemographic**

3. b. The missing value counts are as follows 125, 87, 506, 656, 302 and 87 respectively.

4. In the 'DOB' one customer has a year of birth in '1843',  this obviously is an error

In [8]:
CustomerDemographic.job_industry_category.unique()

array(['Health', 'Financial Services', 'Property', 'IT', nan, 'Retail',
       'Argiculture', 'Manufacturing', 'Telecommunications',
       'Entertainment'], dtype=object)

In [9]:
CustomerDemographic.duplicated().sum()

0

In [10]:
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


**CustomerAddress**

1.There are 3999 non-null values in the address dataset. While there are 4000 unique customers. Customer with customer ID '3' doesn't have data in the address dataset. 

In [11]:
CustomerAddress.head(20)

customer_id                  address  postcode            state  \
0             1       060 Morning Avenue      2016  New South Wales   
1             2      6 Meadow Vale Court      2153  New South Wales   
2             4       0 Holy Cross Court      4211              QLD   
3             5      17979 Del Mar Point      2448  New South Wales   
4             6         9 Oakridge Court      3216              VIC   
5             7         4 Delaware Trail      2210  New South Wales   
6             8      49 Londonderry Lane      2650  New South Wales   
7             9          97736 7th Trail      2023  New South Wales   
8            11     93405 Ludington Park      3044              VIC   
9            12  44339 Golden Leaf Alley      4557              QLD   
10           13      2 Sutherland Street      3799              VIC   
11           14          9 Mcbride Trail      2760  New South Wales   
12           15   9861 New Castle Avenue      2428  New South Wales   
13           16         52 Moland Street      3331              VIC   
14           17    82391 Kensington Lane      3058              VIC   
15           18            092 2nd Alley      2135  New South Wales   
16           19        59 Spaight Circle      2233  New South Wales   
17           20     032 Bartelt Crossing      2444  New South Wales   
18           21          18 Jenna Center      2650  New South Wales   
19           24         3 Cordelia Plaza      2153  New South Wales   

      country  property_valuation  
0   Australia                  10  
1   Australia                  10  
2   Australia                   9  
3   Australia                   4  
4   Australia                   9  
5   Australia                   9  
6   Australia                   4  
7   Australia                  12  
8   Australia                   8  
9   Australia                   4  
10  Australia                   6  
11  Australia                   8  
12  Australia                   9  
13  Australia                   4  
14  Australia                   9  
15  Australia                  12  
16  Australia                   9  
17  Australia                   8  
18  Australia                   7  
19  Australia                  10

In [12]:
CustomerAddress.state.unique()

array(['New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria'], dtype=object)

**CustomerAddress**

2. The state column has 5 unique values namely 'New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria' , but 'New South Wales' and "NSW" are actually the same also 'Victoria' and 'VIC' are also actually the same place. The actual unique values should be 'New South Wales', 'Queensland' and 'Victoria'.

In [13]:
Transcations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [14]:
null_count = []
null_count = Transcations['online_order'].isnull().sum(), Transcations['brand'].isnull().sum(), Transcations['product_line'].isnull().sum(), Transcations['product_class'].isnull().sum(), Transcations['product_size'].isnull().sum(), Transcations['standard_cost'].isnull().sum(), Transcations['product_first_sold_date'].isnull().sum()
null_count

(360, 197, 197, 197, 197, 197, 197)

**Transcations**

1. a. There are missing values in the 'online_order', 'brand', 'product_line', 'product_class', 'product_size', 'standard_cost', 'product_first_sold_date'
2. Explicitly there are 197 rows that have no data on  'brand', 'product_line', 'product_class', 'product_size', 'standard_cost', 'product_first_sold_date'. These are the missing values in those columns. 
3. There are 360 missing values in the 'online_order' column 

In [15]:
Transcations.head(8)

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   
5               6          25         2339       2017-03-08           1.0   
6               7          22         1542       2017-04-21           1.0   
7               8          15         2459       2017-07-15           0.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   
5     Approved  Giant Bicycles         Road        medium       medium   
6     Approved        WeareA2B     Standard        medium       medium   
7     Approved        WeareA2B     Standard        medium       medium   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0  
5     1538.99         829.65                  39031.0  
6       60.34          45.26                  34165.0  
7     1292.84          13.44                  39915.0

**Transcations**

4. 'product_first_sold_date' column doesn't have data that represent what it column name describes.
5. 'Transaction_date' column in format mm/dd/yyyy must be standardise to sync with that of 'DOB' column in both the CustomerDemographics and NewCustomerList tables which is the format yyyy-mm-dd

In [16]:
Transcations.duplicated().sum()

0

In [17]:
NewCustomerList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [18]:
null_count = []
null_count = NewCustomerList['last_name'].isnull().sum(), NewCustomerList['DOB'].isnull().sum(), NewCustomerList['job_title'].isnull().sum(), NewCustomerList['job_industry_category'].isnull().sum(), 
null_count

(29, 17, 106, 165)

**NewCustomerList**

1. There are 29, 17, 106 and 165 missing values in 'last_name', 'DOB', 'job_title'and 'job_industry_category' columns respectively.

In [19]:
NewCustomerList.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.56      0.7000       0.8750     0.743750   
1                  11        0.89      0.8900       1.1125     0.945625   
2                   5        1.01      1.0100       1.0100     1.010000   
3                   1        0.87      1.0875       1.0875     1.087500   
4                   9        0.52      0.5200       0.6500     0.650000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [20]:
NewCustomerList.state.unique()

array(['QLD', 'NSW', 'VIC'], dtype=object)

**NewCustomerList**

2. The 'State' column must be standardised to sync with the data from the CustomerDemographyics table. Hence values 'QLD, 'NSW' and 'VIC' must be changed to full names of the states

3. 'property_valuation', 'past_3_years_bike_related_purchases' and 'postcode' columns have some data stored as text instead of numbers (int or float)

In [21]:
Transcations.brand.unique()


array(['Solex', 'Trek Bicycles', 'OHM Cycles', 'Norco Bicycles',
       'Giant Bicycles', 'WeareA2B', nan], dtype=object)

In [22]:
CustomerAddress.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [23]:
duplicates = CustomerAddress.duplicated(subset=['address'], keep=False)
duplicate_rows = CustomerAddress[duplicates]
duplicate_rows

customer_id                  address  postcode state    country  \
732           737         3 Talisman Place      4811   QLD  Australia   
2315         2320   64 Macpherson Junction      2208   NSW  Australia   
2328         2333  3 Mariners Cove Terrace      3108   VIC  Australia   
2470         2475         3 Talisman Place      4017   QLD  Australia   
2980         2985  3 Mariners Cove Terrace      2216   NSW  Australia   
3535         3540   64 Macpherson Junction      4061   QLD  Australia   

      property_valuation  
732                    2  
2315                  11  
2328                  10  
2470                   5  
2980                  10  
3535                   8